### DISCUSSION TITLE
Access to persisted id in transaction

### DISCUSSION THREAD

**Discussion by Jasenkoo** [mikro-orm/mikro-orm#2488] [Q&A] [answered]
I am using a mikro-orm in a quite large app and it is relying on database queries heavily.

Basically, my project is using nestjs, entity decorators with custom repositories and Postgres database, nothing very special

But I have one problem where it would be nice to have it solved, I couldn't find any in the documentation and I am hoping someone can help me if I am doing something wrong.

Let's say we have a code block in the service with something like this

```ts
async persistMultipleData(data1, data2, data3): Promise<void> {
    this.repostiory1.persist(data1)

    // It would be great if I can get entity id for persistedData2 while in transaction
    const persistedData2 = this.repository2.persist(data2);

    this.repository3.persist(data3);
   
    // Commit transaction
    this.orm.em.flush();
}
```

As you can see in the comment, there is no way to access persisted id while in a transaction. I would need to `commit` and then read again. It would be awesome to do all of that in a single transaction. Is something like that possible with mikro-orm

**Comment by B4nan** [comment, marked as answer]
How do you expect getting the ID without commiting? No flush = no insert queries = no ID exists yet. The ID is accessible on the entity right after the query is fired. 

What do you need that ID for? 

Obvious workaround is to use client side generated PKs, e.g. uuid. Still without making the queries, it is just a value, not an actual PK. 

Another one is to use explicit transaction and flush earlier to get the value.

**Comment by Jasenkoo** [comment]
Hmm, ok, because with TypORM you could wrap the whole code block in the transaction, for example

1. Persist some entity and get id
2. Save the file on disk using the entity id as a reference
3. I saving fails for some reason, the transaction would not commit.

**Reply by B4nan** [reply]
yes, that is what i just suggested to you, use explicit transaction and flush as early as when you need the ID

**Comment by Jasenkoo** [comment]
Ok, tnx for answer
